pkg/providers/aws/provider_postgres_metrics.go

```diff
@@ -0,0 +1,211 @@
+//   * return generic cloud metric data to metric controller to be exposed
+//
+// this provider does not
+//   * expose the metrics, this is controller at a higher level (controller)
+//
+package aws
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/aws/aws-sdk-go/service/cloudwatch/cloudwatchiface"
+	"github.com/integr8ly/cloud-resource-operator/pkg/apis/integreatly/v1alpha1"
+	"github.com/integr8ly/cloud-resource-operator/pkg/providers"
+	"github.com/integr8ly/cloud-resource-operator/pkg/resources"
+	errorUtil "github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"time"
+)
+
+const (
+	postgresMetricProviderName = "aws postgres metrics provider"
+	cloudWatchRDSDBDimension   = "DBInstanceIdentifier"
+)
+
+// builds a list of the metric types and the corresponding metric statistics we want
+var rdsCloudWatchMetricTypes = []providers.CloudProviderMetricType{
+	{
+		PromethuesMetricName: "cro_postgres_free_storage_average",
+		ProviderMetricName:   "FreeStorageSpace",
+		Statistic:            "Average",
+	},
+	{
+		PromethuesMetricName: "cro_postgres_free_storage_average_maximum",
+		ProviderMetricName:   "FreeStorageSpace",
+		Statistic:            "Maximum",
+	},
+	{
+		PromethuesMetricName: "cro_postgres_free_storage_average_minimum",
+		ProviderMetricName:   "FreeStorageSpace",
+		Statistic:            "Minimum",
+	},
+}
+
+var _ providers.MetricsProvider = (*PostgresMetricsProvider)(nil)
+
+type PostgresMetricsProvider struct {
+	Client            client.Client
+	Logger            *logrus.Entry
+	CredentialManager CredentialManager
+	ConfigManager     ConfigManager
+}
+
+func NewAWSPostgresMetricsProvider(client client.Client, logger *logrus.Entry) *PostgresMetricsProvider {
+	return &PostgresMetricsProvider{
+		Client:            client,
+		Logger:            logger.WithFields(logrus.Fields{"providers": postgresMetricProviderName}),
+		CredentialManager: NewCredentialMinterCredentialManager(client),
+		ConfigManager:     NewDefaultConfigMapConfigManager(client),
+	}
+}
+
+func (p *PostgresMetricsProvider) SupportsStrategy(strategy string) bool {
+	return strategy == providers.AWSDeploymentStrategy
+}
+
+// ScrapeMetrics returns scraped metrics to metric controller
+func (p PostgresMetricsProvider) ScrapeMetrics(ctx context.Context, postgresInput *providers.ScrapeMetricsInput) (*providers.ScrapeMetricsOutput, error) {
+	logger := resources.NewActionLoggerWithFields(p.Logger, map[string]interface{}{
+		resources.LoggingKeyAction: "ScrapeMetrics",
+		"Resource":                 postgresInput.Name,
+	})
+	logger.Infof("reconciling postgres metrics %s", postgresInput.Name)
+
+	postgres := &v1alpha1.Postgres{}
+	err := p.Client.Get(ctx, types.NamespacedName{Name: postgresInput.Name, Namespace: postgresInput.Namespace}, postgres)
```

Comment:
Relates to my previous comment. This API call wouldn't be needed if we were implementing an interface like `PostgresMetricsProvider` rather than just `MetricsProvider` because the `ScrapeMetrics` function could accept the Postgres CR